### DISCUSSION TITLE
Hex code (of image) to .png

### DISCUSSION THREAD

**Discussion by Jorg3cSs** [ImageMagick/ImageMagick#4034] [Help]
I have a very large hex code from a 1920x1200 image in a .txt file, all in one line and horizontally, and i want convert that hex to .png, .jpg or .bmp.

So, for example, if i have this hex code in a file.txt:

`FFDDD6FFDDD6A42F28A42F28`

is posible convert to a png, jpg or bmp file for view in my gallery of my phone?

I try different commands, but with no luck.

One command that i try give me a file Test.png, but can't see the image, the image load and load and never show up...

My info:
> Android Version: 10
> Emui Version: 10.0.0
> Termux Version: 0.101
> Imagemagick Version: 7.1.0-4 Q16 aarch64 2021-07-18

**Comment by fmw42** [comment]
Is that code for 8-bit or 16-bit?  What is the desired WxH for the output for the given code string? Is it one row of pixels or how many pixels in a row and how many columns?

**Comment by snibgo** [comment]
IM can't read a text file that has has hex digits all in one long line.

For a list of formats that IM can read, type "magick -list format".

You can use another program to read the file and convert it into a format IM can read.

**Reply by fmw42** [reply]
If you know it is hex represent 8-bits as 6 alpha numeric characters each, you can split the text after every 6th character.

**Reply by Jorg3cSs** [reply]
> IM can't read a text file that has has hex digits all in one long line.
> 
> For a list of formats that IM can read, type "magick -list format".
> 
> You can use another program to read the file and convert it into a format IM can read.

So, i have to convert all this:
rrggbbrrggbbrrggbb
In:
#rrggbb #rrggbb #rrggbb
Or:
#rr gg bb #rr gg bb #rr gg bb
Or:
#rrggbb
#rrggbb
#rrggbb
Save in a .txt or .rgb, (i don't know), and then convert to .bmp, jpg or png?

**Reply by fmw42** [reply]
Which format do you want?

#rrggbb #rrggbb #rrggbb

Or:
#rrggbb
#rrggbb
#rrggbb

**Reply by fmw42** [reply]
This string FFDDD6FFDDD6A42F28A42F28 only has 25 characters. It must be a multiple of 6 (either 24 or 30 for example)

**Reply by Jorg3cSs** [reply]
The string
FFDDD6FFDDD6A42F28A42F28
is only a example. 
Is a simple 2x2 image, are 4 color in hex,
#FFDDD6
#FFDDD6
#A42F28
#A42F28

My question is how to convert that 4 #hex to the original 2x2.jpg image.

**Reply by fmw42** [reply]
I am on Unix-like Mac, so my code would not help you on Windows.

**Reply by Jorg3cSs** [reply]
I'm on android, not windows.

**Reply by fmw42** [reply]
I do not know android coding. Sorry. Just bash Unix shell scripting.

**Comment by snibgo** [comment]
Okay, suppose the file hex2.bin contains:
```
FFDDD6FFDDD6A42F28A42F28
```
... where each pair of hex digits represents rr or gg or bb, how can we convert that to an image with four RGB pixels?

I would start with a trivial C program that read each pair of hex digits and emitted a binary byte, so "ff" became byte-value 255, etc.

Can we do the job entirely within IM? Yes, but it is messy. Windows BAT script:
```
magick ^
  -size 12x2 -depth 8 ^
  gray:hex2.bin ^
  -fx "uu=u*255;uu<58?(uu-48)/255:uu<71?(uu-55)/255:u" ^
  gray:h3.bin

magick ^
  -size 2x12 -depth 8 ^
  gray:h3.bin ^
  -crop 2x@ ^
  +depth ^
  -poly 16,1,1,1 ^
  -rotate -90 ^
  -crop 3x +repage ^
  -append ^
  -crop 3x@ +repage ^
  -combine ^
  out.png
```
I expect this can be done in a single IM command, but I have hit a bug in IM or my brain or both.

out.png contains the required values:
```
magick out.png txt:

# ImageMagick pixel enumeration: 1,4,255,srgb
0,0: (255,221,214)  #FFDDD6  srgb(255,221,214)
0,1: (255,221,214)  #FFDDD6  srgb(255,221,214)
0,2: (164,47,40)  #A42F28  srgb(164,47,40)
0,3: (164,47,40)  #A42F28  srgb(164,47,40)
```
